### ISSUE TITLE
CORS 

### ISSUE THREAD

**Issue by ZenLiuCN** [ktorio/ktor#1656]
**Ktor Version and Engine Used (client or server and name)**
1.3.0

**Describe the bug**
CORS can't pass on some none standard orgin like `chrome-extension://mfgdmpfihlmdekaclngibpjhdebndhdj`

**To Reproduce**
open some new page in chrome or other browser,just call any api,CORS throw would error(brower side).

**Expected behavior**
should be `AnyOrigin` pass by CORS, when we config it as `host("*")` or `anyHost()`

**Comment by cy6erGn0m** [comment]
What exact error? What CORS headers do you see? Does it do preflight request? What are the results of the preflight?

**Comment by ZenLiuCN** [comment]
yes it do preflight request, then it got result as 404 or 403. I will check it tomorrow . maybe it's not bug.

**Comment by ZenLiuCN** [comment]
```kotlin
this.isValidOrigin(call.request.header("chrome-extension://mfgdmpfihlmdekaclngibpjhdebndhdj")!!)
```
this code get `false` , so this `isValidOrigin` can't pass orign like chrome-extension://mfgdmpfihlmdekaclngibpjhdebndhdj, this is confirmed

**Comment by ZenLiuCN** [comment]
raw code is 
```kotlin
		val origin = call.request.headers.getAll(HttpHeaders.Origin)?.singleOrNull()
			//?.takeIf(this::isValidOrigin) //this line comment out is all passed
			?: return
```

**Comment by cy6erGn0m** [comment]
I see, it is caused by a custom protocol `chrome-extension`, need to look deeper into RFCs to see if we should allow or prohibit other protocols.

**Comment by ZenLiuCN** [comment]
ok, Thanks a lot. I current just use a copy of those codes to make it useable.

**Comment by cy6erGn0m** [comment]
The other problem is that it's not clear how to interpret CORS semantics with custom protocols. With a custom protocol, there is neither "host" nor "site" known so it is not clear how do we validate `Origin` header and do further steps. My first assumption is that we should deny any incompatible schemas.

**Comment by ZenLiuCN** [comment]
in some conditions, it may make service usage be limited. Eg. I just use ktor as light service for a chrome extension, and as now I cannot escape the CORS check for it.
BUT, IN Most conditions, this may not a problem. 
And also, my workmate found `http://localhost:3000` seems can't CORS in Firefox Developer vesrion. I'm still checking it.

**Comment by ZenLiuCN** [comment]
uptodate, there is nothing else about ktor CROS.
Hope that this issue will be considered. I really like Ktor, also wanna it could be used in more conditions. Maybe add some switch for this?

**Comment by drampelt** [comment]
Any further thoughts on this? Also running into this problem building a browser extension.

What about requiring the custom protocol to be explicitly included in the `schemes` argument? For example:
```kotlin
install(CORS) {
  host("chrome-extension-id", schemes = listOf("chrome-extension"))
}
```

Then the default scheme will still just be `http` which shouldn't cause any issues that I can think of.

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-469">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by e5l** [comment]
Fixed in Ktor 1.5.3 release
